content/things/js-to-asc.md

```diff
@@ -0,0 +1,563 @@
+
+## Advantages of WebAssembly
+
+In my perception, a lot of people think of WebAssembly purely as a performance primitive. It’s compiled, so it’s gotta be fast, right? Well, for the longest time [I have been vocal that WebAssembly and JavaScript have the same _peak_ performance][io19 talk], and I still stand behind that. Given ideal conditions, they both compile to machine code and end up being equally fast. But there’s obviously more nuance here, and when have conditions _ever_ been ideal on the web‽ Instead, I think it would be better if we thought about WebAssembly as a way to get _predictable_ performance.
+
+However, it’s also important to realize that WebAssembly has recently been getting access to performance primitives (like SIMD or shared-memory threads) that JavaScript cannot utilize, giving WebAssembly a _potential_ edge to predictably out-perform JavaScript. There are also some other qualities of WebAssembly that might make it better suited in specific situations than JavaScript:
+
+### No warmup
+
+When V8 gets given JavaScript, it is first given to the interpreter “Ignition”. It is optimized to make code run as _soon_ as possible. Meanwhile, “Sparkplug” takes Ignitions output (called “byte code”) and turns it into machine code, yielding better performance at the cost of increased memory footprint. While your code is running (either through Ignition or Sparkplug), it is closely observed by V8. It takes note of what kind of type of data you store in your variables, function parameters and so on. Once sufficient data has been collected, V8’s optimizing compiler “TurboFan” kicks in and generates low-level machine code that is optimized for those types. This will give another significant speed boost.
+
+WebAssembly, on the other hand, is strongly typed. It can be turned into machine code _straight away_. V8 has a streaming Wasm compiler called “Liftoff“ which, like Ignition, is geared to get your code running _quickly_, at the cost of generating potentially suboptimal execution speed. The second Liftoff is done, TurboFan kicks in and both generates optimized machine code that will run faster than what Liftoff produced, but will take  longer to generate. The big difference to JavaScript is that the TurboFan can do its work without having to observe your Wasm first.
+
+### No tierdown
+
+The machine code that TurboFan generates for JavaScript is only usable for as long as the assumptions about types hold. If TurboFan generated machine code for a funtion `f` with a number as a parameter, and now all of the sudden that function `f` gets called with an object, the engine has to fall back to Ignition or Sparkplug. That’s called a “deoptimization” (or “deopt” for short). Again, because WebAssembly is strongly typed, the types _can’t_ change. Not only that, but the types that WebAssembly supports are designed to map well to machine code. Deopts can’t happen with WebAssembly.
```

Comment:
typo: "funtion" missing 'c'
Maybe s/all of the sudden/all of a sudden/ ?

Reply:
Thanks!